File: packages/zod/src/claim.ts
import { z } from "zod";
import { ZAuditResponse } from "./audit.js";

export const ZClaimStatus = z.enum([
  "pending",
  "processing",
  "ocr_complete",
  "needs_review",
  "ocr_failed",
  "policy_matched",
  "auditing",
  "approved",
  "flagged",
  "rejected",
]);

export const ZExpenseCategory = z.enum([
  "meals",
  "transport",
  "lodging",
  "other",
]);

export const ZSubmitClaimResponse = z.object({
  claimId: z.string().uuid(),
  status: ZClaimStatus,
  message: z.string(),
});

export const ZAdminClaimFlagFilter = z.enum(["all", "flagged", "unflagged"]);

export const ZAdminClaimDateField = z.enum(["submitted", "claimed"]);

export const ZAdminClaimSortBy = z.enum([
  "submittedDate",
  "claimedDate",
  "amount",
  "status",
  "merchant",
]);

export const ZAdminClaimSortDir = z.enum(["asc", "desc"]);

export const ZAdminClaimListQuery = z.object({
  q: z.string().optional(),
  statuses: z.string().optional(),
  uploaderUserId: z.string().optional(),
  flagged: ZAdminClaimFlagFilter.optional(),
  dateField: ZAdminClaimDateField.optional(),
  dateFrom: z.string().optional(),
  dateTo: z.string().optional(),
  sortBy: ZAdminClaimSortBy.optional(),
  sortDir: ZAdminClaimSortDir.optional(),
});

export const ZClaimResponse = z.object({
  id: z.string().uuid(),
  userId: z.string(),
  receiptFileId: z.string().uuid(),
  businessPurpose: z.string(),
  claimedDate: z.string(),
  expenseCategory: ZExpenseCategory,
  status: ZClaimStatus,
  merchantName: z.string().nullable().optional(),
  receiptDate: z.string().nullable().optional(),
  amount: z.number().nullable().optional(),
  currency: z.string().nullable().optional(),
  dateMismatch: z.boolean(),
  ocrError: z.string().nullable().optional(),
  createdAt: z.string().datetime(),
  updatedAt: z.string().datetime(),
});

export const ZAdminClaimPolicyChunkResponse = z.object({
  chunkText: z.string(),
  category: z.string(),
  pageNum: z.number().int(),
  score: z.number(),
});

export const ZAdminClaimDetailResponse = z.object({
  claim: ZClaimResponse,
  audit: ZAuditResponse.nullable(),
  auditHistory: z.array(ZAuditResponse),
  policyId: z.string().uuid().nullable(),
  policyChunks: z.array(ZAdminClaimPolicyChunkResponse),
});

export type ClaimStatus = z.infer<typeof ZClaimStatus>;
export type ExpenseCategory = z.infer<typeof ZExpenseCategory>;
export type SubmitClaimResponse = z.infer<typeof ZSubmitClaimResponse>;
export type ClaimResponse = z.infer<typeof ZClaimResponse>;
export type AdminClaimPolicyChunkResponse = z.infer<
  typeof ZAdminClaimPolicyChunkResponse
>;
export type AdminClaimDetailResponse = z.infer<
  typeof ZAdminClaimDetailResponse
>;
export type AdminClaimFlagFilter = z.infer<typeof ZAdminClaimFlagFilter>;
export type AdminClaimDateField = z.infer<typeof ZAdminClaimDateField>;
export type AdminClaimSortBy = z.infer<typeof ZAdminClaimSortBy>;
export type AdminClaimSortDir = z.infer<typeof ZAdminClaimSortDir>;
export type AdminClaimListQuery = z.infer<typeof ZAdminClaimListQuery>;

File: apps/backend/internal/model/claim.go
package model

import (
	"time"

	"github.com/google/uuid"
)

type ClaimStatus string
type ExpenseCategory string

const (
	ClaimStatusPending       ClaimStatus = "pending"
	ClaimStatusProcessing    ClaimStatus = "processing"
	ClaimStatusOCRComplete   ClaimStatus = "ocr_complete"
	ClaimStatusNeedsReview   ClaimStatus = "needs_review"
	ClaimStatusOCRFailed     ClaimStatus = "ocr_failed"
	ClaimStatusPolicyMatched ClaimStatus = "policy_matched"
	ClaimStatusAuditing      ClaimStatus = "auditing"
	ClaimStatusApproved      ClaimStatus = "approved"
	ClaimStatusFlagged       ClaimStatus = "flagged"
	ClaimStatusRejected      ClaimStatus = "rejected"
)
const (
	ExpenseCategoryMeals     ExpenseCategory = "meals"
	ExpenseCategoryTransport ExpenseCategory = "transport"
	ExpenseCategoryLodging   ExpenseCategory = "lodging"
	ExpenseCategoryOther     ExpenseCategory = "other"
)

type ReceiptFile struct {
	Base
	FilePath     string `json:"filePath" db:"file_path"`
	OriginalName string `json:"originalName" db:"original_name"`
	MimeType     string `json:"mimeType" db:"mime_type"`
	SizeBytes    int64  `json:"sizeBytes" db:"size_bytes"`
	FileHash     string `json:"fileHash" db:"file_hash"`
	GCSPath      string `json:"gcsPath" db:"gcs_path"`
}

type Claim struct {
	Base
	UserID          string          `json:"userId"          db:"user_id"`
	OrgID           string          `json:"orgId"           db:"org_id"`
	SubmittedByRole string          `json:"submittedByRole" db:"submitted_by_role"`
	ReceiptFileID   uuid.UUID       `json:"receiptFileId"   db:"receipt_file_id"`
	BusinessPurpose string          `json:"businessPurpose" db:"business_purpose"`
	ClaimedDate     time.Time       `json:"claimedDate"     db:"claimed_date"`
	ExpenseCategory ExpenseCategory `json:"expenseCategory" db:"expense_category"`
	Status          ClaimStatus     `json:"status"          db:"status"`
	// OCR-extracted fields — nullable until OCR runs
	MerchantName *string    `json:"merchantName,omitempty" db:"merchant_name"`
	ReceiptDate  *time.Time `json:"receiptDate,omitempty"  db:"receipt_date"`
	Amount       *float64   `json:"amount,omitempty"       db:"amount"`
	Currency     *string    `json:"currency,omitempty"     db:"currency"`
	OCRRawJSON   *string    `json:"ocrRawJson,omitempty"   db:"ocr_raw_json"`
	DateMismatch bool       `json:"dateMismatch"           db:"date_mismatch"`
	OCRError     *string    `json:"ocrError,omitempty"     db:"ocr_error"`

	PolicyID         *uuid.UUID `db:"policy_id"`
	PolicyChunksUsed []byte     `db:"policy_chunks_used"`
}
